### PR TITLE
fix: rank perfect ANOVA separators first

### LINKED ISSUE #147: FClassif scores perfect-separating features 0.0 (ss_within = 0) — SelectKBest drops the most informative feature first
## Description

`FClassif::score` computes the ANOVA F-statistic but guards division with:

```rust
let f_stat = if ss_within > 1e-15 && df_within > 0.0 {
    (ss_between / df_between) / (ss_within / df_within)
} else {
    0.0
};
```

(`src/feature_selection/select_kbest.rs:139-143`)

A feature that **separates the classes perfectly** — each group internally constant, e.g. any deterministic function of the label — has `ss_within == 0` and mathematically `F = ∞`. The implementation assigns `0.0`, and `SelectKBest` sorts scores descending, so a perfect separator is ranked **last** and is the first feature dropped. scikit-learn's `f_classif` returns `inf` in this case and such features rank first — the opposite outcome.

## Location

- `src/feature_selection/select_kbest.rs:139-143`

## Impact

Inverted selection for exactly the features users most want to keep: discretized/binned signals, rule-derived flags, label-leak detectors in sanity-check pipelines.

## Suggested Fix

Return `f64::INFINITY` when `ss_within == 0 && ss_between > 0`; keep `0.0` only when `ss_between == 0` as well (feature carries no class information). Document the convention in the trait docs so custom `ScoreFunction` impls follow it.

Edge: several features scoring `∞` tie — the stable sort keeps input order, making selection deterministic; document this.

## Testing

- Perfect separator selected with k=1
- Constant feature scored 0.0 and dropped
- Mixed finite/infinite scores ordered correctly
- Two perfect separators → deterministic tie-break by input order

## Related Issues

- SelectKBest (#52 SelectPercentile will reuse scoring conventions)

### PR DESCRIPTION
**Pull request by Freakz2z**
## Summary
- score zero-within-class-variance features as `+∞` when they separate classes, while keeping constant features at `0.0`
- compute between-group variance around each feature’s mean rather than the target-label mean
- document infinite-score and stable-tie behavior and add regression coverage for ranking, constants, and ties

## Test plan
- `cargo test`
- `cargo clippy --all-targets --all-features -- -D warnings`
- `cargo fmt -- --check`

Closes #147

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved classification-based feature scoring for perfectly separated classes, decimal-valued data, and partially null datasets.
  * Improved numerical stability for features with large offsets and small real variance.
  * Corrected null-target handling while preserving feature-row alignment.
  * Constant features receive a score of zero, while true perfect separators retain positive-infinity scores.
  * Feature rankings remain consistent for tied infinite scores and near-degenerate datasets.

* **Documentation**
  * Clarified null handling, validation rules, numerical stability, perfect separation, and edge-case scoring behavior.

### CODE REVIEW

**Comment by Freakz2z** [comment]
Addressed the two-sample edge case in 251cd14 by checking zero within-class variance before the zero-degrees-of-freedom fallback. Added a regression test for values [0.0, 1.0] with matching class labels. Full validation passes: 503 unit tests, 8 integration tests, 50 doc tests, rustfmt, and clippy with warnings denied.

**Comment by Freakz2z** [comment]
Addressed the null-handling review in 7a6a2e7. FClassif now derives the mean, observed classes, and both degrees of freedom from each feature's non-null rows. All-null features and features left with one observed target class return actionable errors instead of NaN or an invalid score. Added regressions for the exact partial-null F statistic, downstream SelectKBest ranking, all-null input, and one-class-after-filtering input.

Validation: 507 unit tests, 8 integration tests, 50 doc tests, rustfmt, and clippy with warnings denied all pass.

File: src/feature_selection/select_kbest.rs [hunk 1]
```diff
@@ -95,38 +99,66 @@ impl ScoreFunction for FClassif {
...
+            let n_observed = observed.len() as f64;
+            let n_classes = observed_classes.len() as f64;
             let df_between = n_classes - 1.0;
-            let df_within = n - n_classes;
+            let df_within = n_observed - n_classes;
 
-            let f_stat = if ss_within > 1e-15 && df_within > 0.0 {
-                (ss_between / df_between) / (ss_within / df_within)
-            } else {
+            let f_stat = if ss_within == 0.0 {
```
Comment (DeathSurfing) [inline src/feature_selection/select_kbest.rs:170]: 🔴 **Blocking:** exact float comparison misses rounding residue — verified locally on this branch:

1. `const01 = [0.1, 0.1, 0.1]`, target `[0, 0, 1]` → **inf**. `feature_mean` becomes `0.30000000000000004`, leaving `ss_between ≈ 3.7e-34 > 0`, so a *constant* feature gets ranked first — contradicting "a constant feature scores zero" in the docs above and reintroducing the #147 inversion for non-exact constants. The added test passes only because `5.0` sums exactly.
2. `sep = [0.1, 0.1, 0.1, 0.2, 0.2]`, target `[0, 0, 0, 1, 1]` → **6.23e31**, not inf: `ss_within ≈ 5.8e-34` fails `== 0.0`, so multiple near-perfect separators no longer tie deterministically.

Suggested direction: compare against a data-scale tolerance instead of exact zero, e.g.

```rust
let ss_total: f64 = observed.iter().map(|(v, _)| (v - feature_mean).powi(2)).sum();
let tol = ss_total * f64::EPSILON;
let f_stat = if ss_within <= tol {
    if ss_between > tol { f64::INFINITY } else { 0.0 }
} else if df_within <= 0.0 {
    0.0
} else {
    (ss_between / df_between) / (ss_within / df_within)
};
```

and add regression tests using `0.1`-valued constants and separators at odd group sizes.

**Review by DeathSurfing — CHANGES_REQUESTED** [review]
**Changes requested** — 1 blocking finding (reproduced locally on this branch), 1 repo-convention nit. The core math and null handling are right, and #147 is genuinely fixed for exactly-representable inputs — the exact-zero comparisons just need a scale-relative tolerance.

### 🔴 Blocking
- **src/feature_selection/select_kbest.rs:156** — exact `== 0.0` float checks miss rounding residue. Two repros run locally against this branch:
  1. Constant feature `[0.1, 0.1, 0.1]`, target `[0, 0, 1]` scores **inf** — the docs on this very PR promise a constant scores `0.0`. `SelectKBest` then ranks the junk constant **first**: the same inversion this PR fixes, reintroduced for any constant that doesn't sum exactly in f64 (odd group sizes, `0.1`-style values). The new test passes only because `5.0` is power-of-two-exact.
  2. Perfect separator `[0.1, 0.1, 0.1, 0.2, 0.2]`, target `[0, 0, 0, 1, 1]` scores **6.23e31**, not inf — multiple such separators no longer tie, breaking the documented stable input-order guarantee.

### ⚠️ Convention
- No `CHANGELOG.md` entry under Unreleased. Recent user-facing changes (#132, #131) each add one; `score()` can now return positive infinity, so a `Fixed` entry fits.

### ✅ Looks good
- Between-group variance around the feature mean with observed-row degrees of freedom is the correct ANOVA formulation; I verified the test expectations by hand (partial-null F = 3.0, complete F = 4.5).
- Null handling design is sound: per-column exclusion with hard errors for all-null / single-observed-class columns beats propagating silent garbage.
- All four items from #147's testing checklist are covered; 11/11 module tests pass locally; CI green (linux/macOS, MSRV, lint, security audit).
- Trait and struct docs now document the inf and tie conventions as the issue requested.

(Note: CodeRabbit's inline finding about `df_within` branch ordering was already addressed in commit 251cd14.)

---

**Comment by Freakz2z** [comment]
Addressed the blocking float-comparison issue in 1f5ca29. The zero-sum-of-squares check now uses a scale-relative tolerance based on machine precision, observed row count, and feature squared magnitude. Added the two decimal-valued reproductions from the review, plus a guard that small real variance at a large feature scale remains finite and positive. I also added the requested Unreleased changelog entry.

Validation: 510 unit tests, 8 integration tests, 50 doc tests, rustfmt, clippy with warnings denied, and rustdoc with warnings denied all pass.

**Comment by Freakz2z** [comment]
Refined the numerical fix in efdf3aa after checking the large-offset edge case. Overall and per-class means now use centered accumulation, so mathematically constant decimal groups produce exact zero within-class variance without an offset-scaled tolerance. Added a one-ULP regression at a `1e9` offset to confirm real variance remains finite and positive.

The full validation remains green: 510 unit tests, 8 integration tests, 50 doc tests, rustfmt, clippy with warnings denied, and rustdoc with warnings denied.

**Comment by Freakz2z** [comment]
Kept the full ANOVA calculation in centered coordinates in c1b6d43. Overall and group means remain offsets, `ss_between` compares those offsets, and `ss_within` uses group-relative deltas, avoiding half-ULP loss when reconstructing large absolute means. Added the successive-representable-value regression at `1e16` and pinned the expected F-score to `2.0`.

Validation: 511 unit tests, 8 integration tests, 50 doc tests, rustfmt, clippy with warnings denied, and rustdoc with warnings denied all pass.

**Comment by Freakz2z** [comment]
@coderabbitai review

**Comment by Freakz2z** [comment]
Addressed target-null row alignment in b5b87ec. `y_vals` now retains its physical `Option<f64>` rows, row-count validation compares physical lengths, and each feature includes a row only when both feature and target values are present. Added regressions for an aligned null-target row containing an extreme feature value and for a physical target-length mismatch; updated the docs and changelog accordingly.

Validation: 513 unit tests, 8 integration tests, 50 doc tests, rustfmt, clippy with warnings denied, and rustdoc with warnings denied all pass.

**Comment by Freakz2z** [comment]
@coderabbitai review

File: src/feature_selection/select_kbest.rs (hunk 1 above)
Comment (Freakz2z) [reply src/feature_selection/select_kbest.rs:170]: Addressed in 1f5ca29, efdf3aa, and c1b6d43. The ANOVA calculation now stays in centered coordinates for both overall and group means, so decimal constants produce exact zero within-class and between-class variance without an offset-scaled tolerance, while real one-ULP variance at a large offset is preserved. Regressions cover the two decimal cases from this thread, one-ULP variance at 1e9, and successive representable values at 1e16 (finite F = 2.0). The latest full validation is 513 unit, 8 integration, and 50 doc tests, plus rustfmt, clippy -D warnings, and rustdoc -D warnings.

**Review by DeathSurfing — APPROVED** [review]
**Approved** — blocking findings resolved. Verified locally against b5b87ec:

- Decimal constant `[0.1, 0.1, 0.1]` scores exactly `0.0`; decimal separator `[0.1 x3, 0.2 x2]` scores exactly `inf` — centered accumulation keeps ss_within/ss_between at exact zero without any tolerance hack.
- Half-ULP means at 1e16 return F = 2.0; one-ULP variance at 1e9 stays finite and positive.
- Null targets keep physical row alignment (count check on `Vec<Option<f64>>`, paired filtering in `observed`).
- CHANGELOG entry present under Unreleased → Fixed.

Local gates: fmt, clippy -D warnings, 516 unit + 8 integration + 50 doc tests, rustdoc -D warnings — all green. Main-merge conflicts resolved identically via 4c5150e.

---
*Review from DeathSurfing*
